### PR TITLE
perf(gemma4): capture decode steps in bucketed CUDA graphs

### LINKED ISSUE #910: gemma4: every decode step pays its launch overhead on the host
## Description

A decode step is 48 decoder layers of small kernels launched one at a time, every token. The work per kernel is fixed and the pointers are stable — the arena change made every buffer, plan and metadata slot persistent and refilled in place — so the host-side launch stream is pure overhead that repeats identically each step.

The eager gate in the engine and the CLI rule that rejects `--cuda-graph=true` are placeholders from bring-up; the graph-stable plan refill path this needs has been in the kernels crate all along.

## Proposed Plan

- Capture the decode step's GPU body — embedding through the LM head, a pure kernel sequence with no allocation and no synchronization — as one CUDA graph per power-of-two batch bucket. A step pads to its bucket with rows that write the pools' reserved padding pages at position zero, never a real request's KV; pad logits rows are never read.
- Make padding the numeric contract: a step computes at its bucket whether or not graphs are on, so a captured replay and the eager escape hatch are the same arithmetic at every batch size, power of two or not.
- Sweep every bucket at startup from one dummy request — one eager warm pass to force lazy CUDA and cuBLAS initialization outside capture, a capture without execution, one replay through the serving path — then synchronize, so capture cost and any capture error land before ready. Serving itself only replays: a bucket the sweep missed is an error, not a lazy capture.
- Flip the contract: `--cuda-graph` defaults to true and explicit false is the eager escape hatch, replacing the rule that rejected it.

## Acceptance Criteria

- The captured replay is bit-identical to the eager path at a batch that is not a power of two, exercising the pad row: the ragged batch determinism gate runs three real requests in a four-row arena, and replay, per-step row moves and the eager comparison all agree per request.
- The serving behaviour checks hold on the graph path: row to request mapping, retirement, queueing, cancellation, window crossing.
- The A/B against the eager base and the resident cost are recorded.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #910

A decode step is 48 decoder layers of small kernels launched one at a time, every token, over pointers that no longer move: the arena change made every buffer, plan and metadata slot persistent and refilled in place. The launch stream is therefore pure per-step overhead, and the step's GPU body — embedding through the LM head, no allocation, no synchronization — is capturable as it stands.

The step is now captured as one CUDA graph per power-of-two batch bucket. A step pads to its bucket with rows that write the pools' reserved padding pages at position zero — never a real request's KV — and pad logits rows are never read; per-step change rides the plan and metadata contents uploaded before launch. Padding is the numeric contract: a step computes at its bucket whether or not graphs are on, so a captured replay and the eager escape hatch are the same arithmetic at every batch size.

A startup sweep drives every bucket from one dummy request — one eager warm pass to force lazy CUDA and cuBLAS initialization outside capture, a capture without execution, then one replay through the serving path — and synchronizes at the end, so capture cost and any capture error land before ready. Serving itself only replays; a bucket the sweep missed is an error rather than a lazy capture on a live request. `--cuda-graph` now defaults to true for this line and explicit false is the eager escape hatch, replacing the bring-up rule that rejected it.

### Test Env

Single GPU (sm_89, x86_64), CUDA 12.9, pinned 12B checkpoint.

### Verification

- fmt; `cargo build`/`clippy -D warnings` for `pegainfer-kernels` and for `pegainfer-gemma4 --features gemma4`; the CI package set with `--locked`; `pegainfer-gemma4 --lib` 26 passed, 9 ignored.
- The checkpoint oracles against the pinned 12B all pass; they drive the eager prefill path, which is unchanged.
- The ragged batch determinism gate passes with graphs on: three real requests in a four-row arena, so the pad row and its reserved-page writes are in every step; replay, per-step row moves, and eager-versus-graph all bit-identical per request at a non-power-of-two batch. On this fixture the captured replay is bit-identical to the eager path.
- The serving behaviour checks pass on the graph path with zero failed requests: row to request mapping, a batch repeated twice byte-identical, in-batch retirement, queueing past the slot ceiling, mid-stream cancellation, and window crossing under concurrency.
- A three-arm A/B, same box, same request set, three runs per arm and per width. The middle arm runs this change with `--cuda-graph=false`, isolating the padding contract's cost from the graphs' gain; c9 pads to the 16 bucket, the worst boundary:

  | | base eager | this change, eager | this change, graphs (default) |
  | --- | --- | --- | --- |
  | single stream TPOT, median | 28.04 ms | 28.30 ms | 27.53 ms |
  | one request | 33.2 tok/s | 33.0 tok/s | 34.0 tok/s |
  | eight requests | 133.9 tok/s | 134.0 tok/s | 137.3 tok/s |
  | nine requests, padded to 16 | 139.2 tok/s | 138.8 tok/s | 141.9 tok/s |
  | ready wall, 1 s resolution | 5 s | 4 s | 5 s |
  | resident idle, nvidia-smi | 32926 MiB | 32926 MiB | 33002 MiB |

  Padding alone costs at most 0.3 percent at the worst boundary and nothing at aligned widths; with graphs on, the default is net positive at every measured width, including the padded one (c9 +1.9 percent over base). The startup sweep is invisible at 1 second resolution: all three arms are ready in 4 to 5 seconds. Under the serving checks the graph-on process peaked at 33386 MiB.

### CODE REVIEW

File: pegainfer-gemma4/src/serve.rs [hunk 1]
```diff
@@ -883,37 +910,187 @@ impl GemmaServe {
...
+                        ids,
+                        padded,
+                        local_plan,
+                        global_plan,
+                        local_origins,
+                        head_normed,
+                        logits,
+                    )
+                })?;
+                self.decode_batch_step(ctx, arena, &mut kvs, &[0])?;
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply pegainfer-gemma4/src/serve.rs:1088]: Not a race. Both the upload and the subsequent launch are enqueued on the same stream (active_cu_stream(ctx); the engine is single-stream and the step arena asserts stream identity). Per the CUDA driver API docs, both operations are stream-ordered: cuGraphUpload — "Each upload is ordered behind both any previous work in hStream and any previous launches of hGraphExec" — and cuGraphLaunch carries the identical ordering guarantee, so a launch enqueued after the upload on the same stream is serialized behind it by the driver. There is also no documented failure mode for launching a partially uploaded exec; launching a never-uploaded exec is legal (the upload happens implicitly). The "synchronize before launching" phrase came from a stale doc comment on capture_only itself, which this revision fixes to state the stream-ordering contract; a synchronize is only needed if the first launch were issued from a different stream, which this engine never does.

**Comment by FeathBow** [comment]
@codex review plz
